### DISCUSSION TITLE
Is there any benchmarks or performance comparison with JDBC?

### DISCUSSION THREAD

**Discussion by orchestr7** [pgjdbc/r2dbc-postgresql#607] [Q&A]
Hi! I am curious is there any official (or not) comparison of performance of RD2BC vs JDBC for postgre?

Also especially interested in the most common scenario with Spring MVC/Webflux.

PS second bonus question, that is more technical - is there any sense (in a relation to performance) to use r2dbc in Java21 or performance problems will be covered with JDBC + virtual threads? 

Will appreciate any thoughts, thank you 🙏❤️

**Comment by mp911de** [comment]
This project contains a few JMH-based benchmarks to compare JDBC to R2DBC throughput performance. Clearly, a proper benchmark isn't just throughput but needs to take resource consumption, threading, and application behavior into account. We do not maintain such an extensive benchmark suite here. You should rather benchmark your own application to learn whether the benefits outweigh complexity introduced by reactive programming paradigms.

If you're doing reactive just because of non-blocking, then you're doing it wrong, and switching to Virtual Threads will retain debugging convenience coupled with the imperative code style.

If you're doing reactive because your application benefits from streaming and backpressure, then you should consider R2DBC.
